### PR TITLE
sonic-snmpagent: fixed interface speeds

### PR DESCRIPTION
**Pull request by MichelMoriniaux**
* ifSpeed now reports speed under 2^32 bps
* ifSpeed returns 2^32 is speed is above
* ifHighSpeed reports the interface speed in Mbps
* ifHighSpeed defaults to 40000 if speed not available in APPL_DB
* unit tests for rfc1213 and rfc2863
* appl_db.json modified to provide variety of speeds

  Signed-off-by: michel.moriniaux@gmail.com

After building code I saw that there was already a pull request for this feature. As there was no activity since February on it, my pull has unit tests and more completeness I went ahead with this one.

### CODE REVIEW

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 1]
```diff
@@ -336,6 +336,19 @@ def get_mtu(self, sub_id):
 
         return int(entry.get(b"mtu", 0))
 
+    def get_speed(self, sub_id):
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: >get_speed [](start = 8, length = 9)

get_speed_bps

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 2]
```diff
@@ -336,6 +336,19 @@ def get_mtu(self, sub_id):
...
+        :param sub_id: The 1-based sub-identifier query.
+        :return: min of 4294967295 or speed value for the respective sub_id.
+        """
+        entry = self._get_if_entry(sub_id)
+        if not entry:
+            return
+
+        speed = int(entry.get(b"speed", 0))
+        # speed is reported in Mbps in the db 
+        return min(4294967295, speed * 1000000)
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: suggest replace the magic number with a constant expression

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py [hunk 3]
```diff
@@ -182,6 +182,34 @@ def _get_counter(self, oid, table_name, mask):
...
+    def _get_if_entry(self, sub_id):
+        """
+        :param oid: The 1-based sub-identifier query.
+        :return: the DB entry for the respective sub_id.
+        """
+        oid = self.get_oid(sub_id)
+        if not oid:
+            return
+
+        table = ""
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc2863.py]: >table [](start = 8, length = 5)

table is not a good name.

Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc2863.py]: >"" [](start = 16, length = 2)

None

File: tests/test_hc_interfaces.py [hunk 4]
```diff
@@ -71,4 +71,67 @@ def test_get_next3(self):
...
+            header=PDUHeader(1, PduTypes.GET, 16, 0, 42, 0, 0, 0),
+            oids=[oid]
+        )
+
+        encoded = get_pdu.encode()
+        response = get_pdu.make_response(self.lut)
+        print(response)
+
+        n = len(response.values)
+        # self.assertEqual(n, 7)
```
Comment (qiluo-msft) [inline tests/test_hc_interfaces.py]: >#  [](start = 8, length = 2)

Do you still need it? If not, remove

File: tests/test_hc_interfaces.py [hunk 5]
```diff
@@ -71,4 +71,67 @@ def test_get_next3(self):
...
```
Comment (qiluo-msft) [inline tests/test_hc_interfaces.py]: >#  [](start = 8, length = 2)

The same

File: tests/test_hc_interfaces.py (hunk 4 above)
Comment (MichelMoriniaux) [reply tests/test_hc_interfaces.py]: there for completeness

File: tests/test_hc_interfaces.py (hunk 5 above)
Comment (MichelMoriniaux) [reply tests/test_hc_interfaces.py]: also for completeness

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py (hunk 3 above)
Comment (MichelMoriniaux) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: I did not write this function, I lifted it directly from rfc1213.py
if we are going to refactor existing utility code I feel it should be done in another pull request and work with the "refactor rfc1213.py and rfc2863.py" issue

Comment (MichelMoriniaux) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: see above comment

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py (hunk 2 above)
Comment (MichelMoriniaux) [reply src/sonic_ax_impl/mibs/ietf/rfc1213.py]: will do

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py (hunk 1 above)
Comment (MichelMoriniaux) [reply src/sonic_ax_impl/mibs/ietf/rfc1213.py]: ok

**Comment by MichelMoriniaux** [comment]
changes done

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py (hunk 3 above)
Comment (qiluo-msft) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: I did not intend to blame you or your code. Even it is copied from the old files, it is your responsibility to make it best at new places.

You may prefer don't touch unrelated code. I am ok. Just keep new code better.

---
In reply to: [207958146](https://github.com/Azure/sonic-snmpagent/pull/73#discussion_r207958146) [](ancestors = 207958146)

File: tests/test_hc_interfaces.py (hunk 4 above)
Comment (MichelMoriniaux) [reply tests/test_hc_interfaces.py]: now I understand you were refering only to the n = len...
I agree, removed

File: tests/test_hc_interfaces.py (hunk 5 above)
Comment (MichelMoriniaux) [reply tests/test_hc_interfaces.py]: same

**Pull request merged**